c7n/logs_support.py

```diff
@@ -0,0 +1,173 @@
+from dateutil import parser
+from dateutil import tz
+from gzip import GzipFile
+
+from c7n.executor import ThreadPoolExecutor
+from c7n.utils import local_session
+
+
+log = logging.getLogger('custodian.logs')
+
+
+def _timestamp_from_string(date_text):
+    try:
+        date_dt = parser.parse(date_text)
+        date_ts = time.mktime(date_dt.timetuple())
+        return long(date_ts * 1000)
+    except (AttributeError, TypeError, ValueError):
+        return 0
+
+
+def normalized_log_entries(raw_entries):
+    '''Mimic the format returned by LambdaManager.logs()'''
+    entry_start = '([0-9:, \-]+) - .* - (\w+) - (.*)$'
+    entry = None
+    # process start/end here - avoid parsing log entries twice
+    for line in raw_entries:
+        m = re.match(entry_start, line)
+        if m:
+            # this is the start of a new entry
+            # spit out the one previously built up (if any)
+            if entry is not None:
+                yield entry
+            (log_time, log_level, log_text) = m.groups()
+            # convert time
+            log_timestamp = _timestamp_from_string(log_time)
+            # join level and first line of message
+            msg = '[{}] {}'.format(log_level, log_text)
+            entry = {
+                'timestamp': log_timestamp,
+                'message': msg,
+            }
+        else:
+            # additional line(s) for entry (i.e. stack trace)
+            entry['message'] = entry['message'] + line
+    if entry is not None:
+        # return the final entry
+        yield entry
+
+
+def log_entries_in_range(entries, start, end):
+    '''filter out entries before start and after end'''
+    start = _timestamp_from_string(start)
+    end = _timestamp_from_string(end)
+    for entry in entries:
+        log_timestamp = entry.get('timestamp', 0)
+        if log_timestamp >= start and log_timestamp <= end:
+            yield entry
+
+
+def log_entries_from_s3(session_factory, output, start, end):
+    client = local_session(session_factory).client('s3')
+    key_prefix = output.key_prefix.strip('/')
+    local_tz = tz.tzlocal()
+    start = datetime.fromtimestamp(
+        _timestamp_from_string(start) / 1000
+    )
+    end = datetime.fromtimestamp(
+        _timestamp_from_string(end) / 1000
+    ).replace(tzinfo=local_tz)
+    records = []
+    key_count = 0
+    log_filename = 'custodian-run.log.gz'
+    marker = '{}/{}/{}'.format(
+        key_prefix,
+        start.strftime('%Y/%m/%d/00'),
+        log_filename,
+    )
+    p = client.get_paginator('list_objects').paginate(
+        Bucket=output.bucket,
+        Prefix=key_prefix + '/',
+        Marker=marker,
+    )
+    with ThreadPoolExecutor(max_workers=20) as w:
+        for key_set in p:
+            if 'Contents' not in key_set:
+                continue
+            keys = [k for k in key_set['Contents']
+                    if k['Key'].endswith(log_filename)
+                    and k['LastModified'].astimezone(local_tz) < end]
```

Comment:
rather than converting every record to local tz, why not convert the local time to utc for comparison, from a number of conversions as cost perspective, one conversion is better than many.